app/src/main/java/com/bnyro/wallpaper/db/dao/FavoritesDao.kt

```diff
@@ -13,21 +13,50 @@ interface FavoritesDao {
     @Query("SELECT * FROM favorites")
     fun getAll(): List<Wallpaper>
 
-    @Query("SELECT * FROM favorites")
-    fun getAllFlow(): Flow<List<Wallpaper>>
+    @Query("SELECT * FROM favorites WHERE favorite = 1")
+    fun getFavorites(): List<Wallpaper>
+
+    @Query("SELECT * FROM favorites WHERE favorite = 1 ORDER BY timeAdded DESC")
+    fun getFavoritesFlow(): Flow<List<Wallpaper>>
 
-    @Query("SELECT * FROM favorites WHERE imgSrc LIKE :imgSrc")
-    fun findBySrc(imgSrc: String): Wallpaper
+    @Query("SELECT * FROM favorites WHERE inHistory = 1 ORDER BY timeAdded DESC")
+    fun getHistoryFlow(): Flow<List<Wallpaper>>
 
-    @Query("SELECT EXISTS (SELECT 1 FROM favorites WHERE imgSrc = :imgSrc)")
-    fun exists(imgSrc: String): Boolean
+    @Query("SELECT EXISTS (SELECT 1 FROM favorites WHERE favorite = 1 AND imgSrc = :imgSrc)")
+    fun isLiked(imgSrc: String): Boolean
 
+    @Query("SELECT * FROM favorites WHERE imgSrc = :imgSrc")
+    fun findByImgSrc(imgSrc: String): Wallpaper?
+
+    /**
+     * Do not use this method directly unless when restoring backups
+     */
     @Insert(onConflict = OnConflictStrategy.REPLACE)
-    fun insertAll(vararg wallpaper: Wallpaper)
+    fun insertAll(wallpapers: List<Wallpaper>)
+
+    fun insert(wallpaper: Wallpaper, isFavorite: Boolean?, isHistory: Boolean?) {
+        val existingWallpaper = findByImgSrc(wallpaper.imgSrc)
+
+        wallpaper.favorite = isFavorite ?: existingWallpaper?.favorite ?: false
+        wallpaper.inHistory = isHistory ?: existingWallpaper?.inHistory ?: false
+
+        if (existingWallpaper != null) {
+            updateWallpaper(wallpaper)
+        } else {
+            wallpaper.timeAdded = System.currentTimeMillis()
+            insertAll(listOf(wallpaper))
+        }
+    }
```

Comment:
Why not just use
```suggestion
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(wallpaper: Wallpaper)
```

If the reason behind this is to update the `timeAdded` value, then it will be better to use
```
@ColumnInfo(defaultValue = "CURRENT_TIMESTAMP") var timeAdded: Long = 0
```

Reply:
Your suggestion with `@Insert` will overwrite the old value of `isFavorite` if the function is called with `isFavorite = null` (because the app has no knowledge about the database stats of this wallpaper at that time, the `Wallpaper` object is exactly the one returned by the API, without any modifications), and same for `isHistory` if `isHistory = null`.

> If the reason behind this is to update the timeAdded value
>
It's not, that timestamp should only be set when you apply / save it the first time.